core/src/main/java/io/questdb/cairo/sql/PageFrameMemoryPool.java

```diff
@@ -39,13 +50,63 @@
  * This pool is thread-unsafe as it may hold navigated Parquet partition data,
  * so it shouldn't be shared between multiple threads.
  */
-// TODO: add LRU cache for multiple frames
-public class PageFrameMemoryPool implements QuietCloseable {
-    private final PageFrameMemoryImpl frameMemory = new PageFrameMemoryImpl();
+public class PageFrameMemoryPool implements QuietCloseable, Mutable {
+    private static final byte FRAME_MEMORY_MASK = 1 << 2;
+    private static final byte RECORD_A_MASK = 1;
+    private static final byte RECORD_B_MASK = 1 << 1;
+    // LRU cache (most recently used buffers are to the right)
+    private final ObjList<ParquetBuffers> cachedParquetBuffers;
+    private final PageFrameMemoryImpl frameMemory;
+    private final ObjList<ParquetBuffers> freeParquetBuffers;
+    // Contains parquet to query column index mapping.
+    private final IntList fromParquetColumnIndexes;
+    private final int parquetCacheSize;
+    // Contains [parquet_column_index, column_type] pairs.
+    private final DirectIntList parquetColumns;
+    private final PartitionDecoder parquetDecoder;
+    // Contains table reader to parquet column index mapping.
+    private final IntList toParquetColumnIndexes;
     private PageFrameAddressCache addressCache;
 
+    public PageFrameMemoryPool(int parquetCacheSize) {
+        try {
+            this.parquetCacheSize = parquetCacheSize;
+            cachedParquetBuffers = new ObjList<>(parquetCacheSize);
+            freeParquetBuffers = new ObjList<>(parquetCacheSize);
+            for (int i = 0; i < parquetCacheSize; i++) {
+                freeParquetBuffers.add(new ParquetBuffers());
+            }
+            frameMemory = new PageFrameMemoryImpl();
+            toParquetColumnIndexes = new IntList(16);
+            fromParquetColumnIndexes = new IntList(16);
+            parquetColumns = new DirectIntList(32, MemoryTag.NATIVE_DEFAULT);
+            parquetDecoder = new PartitionDecoder();
+        } catch (Throwable th) {
+            close();
+            throw th;
+        }
+    }
+
+    @Override
+    public void clear() {
+        Misc.free(parquetDecoder);
+        toParquetColumnIndexes.restoreInitialCapacity();
+        fromParquetColumnIndexes.restoreInitialCapacity();
+        parquetColumns.resetCapacity();
+        freeParquetBuffers.addAll(cachedParquetBuffers);
+        cachedParquetBuffers.clear();
+        Misc.freeObjListAndKeepObjects(freeParquetBuffers);
+        frameMemory.clear();
+        addressCache = null;
+    }
+
     @Override
     public void close() {
+        Misc.free(parquetDecoder);
+        Misc.free(parquetColumns);
+        freeParquetBuffers.addAll(cachedParquetBuffers);
+        cachedParquetBuffers.clear();
+        Misc.freeObjListAndKeepObjects(freeParquetBuffers);
         frameMemory.clear();
```

Comment:
not a new code, but lets keep close() and clear() semantics "clear" :)

Reply:
Removed this call in bff3be12fc4b4bd622676d07b92abeb45962585b